src/parser/docparser.hpp

```diff
@@ -0,0 +1,75 @@
+#pragma once
+#include <nlohmann/json.hpp>
+#include <array>
+
+namespace docgen {
+
+// Tags indicating possible states of a DocParser object.
+enum class State {
+    read = 0,
+    slash,
+    single_line,
+    block,
+    last = block // update this variable if last enum changes
+};
+
+namespace details {
+
+// forward declaration
+// Parse function depending on a particular state.
+template <State state>
+inline const char *parse(const char *begin, 
+                         const char *end,
+                         nlohmann::json& json);
+
+
+} // namespace details
+
+//////////////////////////////////////////////////////
+// DocParser Declaration
+//////////////////////////////////////////////////////
+
+// DocParser is a class that will parse a chunk of text into JSON format.
+class DocParser
+{
+public:
+
+    // Main function that user will call to parse a chunk of text.
+    // This function will update the json object with information extracted from chunk.
+    void parse(const char *begin, 
+               const char *end,
+               nlohmann::json& json);
+
+private:
+    using parse_fn_type = std::function<
+        const char *(const char *, const char *, nlohmann::json&)
+        >;
+    static constexpr size_t array_size = static_cast<size_t>(State::last);
```

Comment:
yes I agree that you do not need static_cast with plain enum. But ok we can just do enum.

Reply:
Ok, I'm still not convinced. What do we not have with `enum class` that we do with `enum`? It's not like we can't go from enum class value to its underlying integer value. We can also go the other way. We just have the extra feature of scoping and no implicit conversion from `enum` to `int`.

Reply:
I really really prefer explicit over implicit when it comes to conversion. It's just a safer way of coding. It's like having that implicit casting away constness warning you get when you pass a const& object to a function that expects regular reference. Like I don't understand why we can't have this extra precaution? You don't want to flood the code with a bunch of `static_cast`s?

Reply:
Ok, because I decided to put the `enum` inside the class, it's scoped enough that I'm comfortable with just `enum`. In the future, if we ever need to have an `enum` in some relatively large namespace, we must use `enum class` whether or not we intend to use them as indices or whatever. First and foremost, we need some scope safety.